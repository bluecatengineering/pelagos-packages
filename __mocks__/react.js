module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
