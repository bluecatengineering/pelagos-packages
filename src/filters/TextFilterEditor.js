import React, {useState, useCallback} from 'react';
import PropTypes from 'prop-types';
import identity from 'lodash-es/identity';

import ListInput from '../listInput/ListInput';
import renderSuggestion from '../suggestions/renderSuggestion';
import __ from '../strings';

const getName = (o) => o.name;
const getSuggestionHighlightKey = (s) => (s.order === 2 ? s.name : null);

/** Filter editor where values are typed. */
const TextFilterEditor = ({label, placeholder, list, getSuggestions, parseInput, validateSaveRef, onListChange}) => {
	const [text, setText] = useState('');
	const [error, setError] = useState(null);

	validateSaveRef.current = useCallback(() => {
		if (text) {
			setError(__('ITEM_NOT_ADDED'));
			return false;
		}
		return true;
	}, [text]);

	return (
		<ListInput
			id="textFilterEditorInput"
			label={label}
			placeholder={placeholder}
			type="text"
			column
			value={text}
			list={list}
			error={error}
			getSuggestions={getSuggestions}
			getSuggestionText={getName}
			getHighlightKey={getSuggestionHighlightKey}
			renderSuggestion={renderSuggestion}
			getItemKey={identity}
			getItemName={identity}
			parseInput={parseInput}
			onListChange={onListChange}
			onTextChange={setText}
			onErrorChange={setError}
		/>
	);
};

TextFilterEditor.propTypes = {
	/** The field label. */
	label: PropTypes.string,
	/** The field hint. */
	placeholder: PropTypes.string,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning suggestions. */
	getSuggestions: PropTypes.func,
	/** Function to parse the input text. */
	parseInput: PropTypes.func,
	/** A ref where the validation function will be stored. */
	validateSaveRef: PropTypes.object,
	/** Function invoked to apply changes. */
	onListChange: PropTypes.func,
};

export default TextFilterEditor;
