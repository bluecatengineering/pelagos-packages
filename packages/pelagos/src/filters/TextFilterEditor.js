import {useState, useCallback} from 'react';
import PropTypes from 'prop-types';
import identity from 'lodash-es/identity';
import {t} from '@bluecateng/l10n.macro';

import ListInput from '../listInput/ListInput';
import renderSuggestion from '../suggestions/renderSuggestion';
import useRandomId from '../hooks/useRandomId';

import FilterEditor from './FilterEditor';

const getName = (o) => o.name;
const getSuggestionHighlightKey = (s) => (s.order === 2 ? s.name : null);

/** Filter editor where values are typed. */
export const OldTextFilterEditor = ({
	label,
	placeholder,
	list,
	getSuggestions,
	parseInput,
	validateSaveRef,
	onListChange,
}) => {
	const [text, setText] = useState('');
	const [error, setError] = useState(null);

	const handleListChange = useCallback((list) => onListChange(list.length ? list : null), [onListChange]);

	validateSaveRef.current = useCallback(() => {
		if (text) {
			setError(t`Press Enter to add items to the list.`);
			return false;
		}
		return true;
	}, [text]);

	return (
		<ListInput
			id="textFilterEditorInput"
			label={label}
			placeholder={placeholder}
			type="text"
			column
			value={text}
			list={list || []}
			error={error}
			getSuggestions={getSuggestions}
			getSuggestionText={getName}
			getHighlightKey={getSuggestionHighlightKey}
			renderSuggestion={renderSuggestion}
			getItemKey={identity}
			getItemName={identity}
			parseInput={parseInput}
			onListChange={handleListChange}
			onTextChange={setText}
			onErrorChange={setError}
		/>
	);
};

OldTextFilterEditor.propTypes = {
	/** The field label. */
	label: PropTypes.string,
	/** The field hint. */
	placeholder: PropTypes.string,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning suggestions. */
	getSuggestions: PropTypes.func,
	/** Function to parse the input text. */
	parseInput: PropTypes.func,
	/** A ref where the validation function will be stored. */
	validateSaveRef: PropTypes.object,
	/** Function invoked to apply changes. */
	onListChange: PropTypes.func,
};

/** Filter editor where values are typed. */
export const NewTextFilterEditor = ({
	id,
	label,
	placeholder,
	list: initialList,
	getSuggestions,
	parseInput,
	chipId,
	onClose,
	onSave,
}) => {
	const [list, setList] = useState(initialList || []);
	const [text, setText] = useState('');
	const [error, setError] = useState(null);

	const handleSave = useCallback(
		() => (text ? setError(t`Press Enter to add items to the list.`) : onSave(list.length ? list : null)),
		[list, onSave, text]
	);

	id = useRandomId(id);
	return (
		<FilterEditor id={id} chipId={chipId} aria-labelledby={`${id}-list-label`} onClose={onClose} onSave={handleSave}>
			<ListInput
				id={`${id}-list`}
				label={label}
				placeholder={placeholder}
				type="text"
				column
				value={text}
				list={list}
				error={error}
				getSuggestions={getSuggestions}
				getSuggestionText={getName}
				getHighlightKey={getSuggestionHighlightKey}
				renderSuggestion={renderSuggestion}
				getItemKey={identity}
				getItemName={identity}
				parseInput={parseInput}
				onListChange={setList}
				onTextChange={setText}
				onErrorChange={setError}
			/>
		</FilterEditor>
	);
};

NewTextFilterEditor.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The field label. */
	label: PropTypes.string,
	/** The field hint. */
	placeholder: PropTypes.string,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning suggestions. */
	getSuggestions: PropTypes.func,
	/** Function to parse the input text. */
	parseInput: PropTypes.func,
	/** The ID of the chip which triggered the editor. */
	chipId: PropTypes.string,
	/** Function invoked to close the editor. */
	onClose: PropTypes.func,
	/** Function invoked to save changes. */
	onSave: PropTypes.func,
};

/** Filter editor where values are typed. */
const TextFilterEditor = ({forArea, ...props}) =>
	forArea ? <NewTextFilterEditor {...props} /> : <OldTextFilterEditor {...props} />;

TextFilterEditor.propTypes = {
	/** The component id (only when forArea is true). */
	id: PropTypes.string,
	/** The field label. */
	label: PropTypes.string,
	/** The field hint. */
	placeholder: PropTypes.string,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning suggestions. */
	getSuggestions: PropTypes.func,
	/** Function to parse the input text. */
	parseInput: PropTypes.func,
	/** A ref where the validation function will be stored (only when forArea is false). */
	validateSaveRef: PropTypes.object,
	/** Function invoked to apply changes (only when forArea is false). */
	onListChange: PropTypes.func,
	/** The ID of the chip which triggered the editor (only when forArea is true). */
	chipId: PropTypes.string,
	/** Function invoked to close the editor (only when forArea is true). */
	onClose: PropTypes.func,
	/** Function invoked to save changes (only when forArea is true). */
	onSave: PropTypes.func,
	/** Whether to enable new behaviour for FilterArea.  */
	forArea: PropTypes.bool,
};

export default TextFilterEditor;
