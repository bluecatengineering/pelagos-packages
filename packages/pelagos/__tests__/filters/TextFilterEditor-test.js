import {useState} from 'react';
import {shallow} from 'enzyme';

import TextFilterEditor, {OldTextFilterEditor, NewTextFilterEditor} from '../../src/filters/TextFilterEditor';

jest.unmock('../../src/filters/TextFilterEditor');

describe('TextFilterEditor', () => {
	describe('rendering', () => {
		it('renders expected elements when forArea is not set', () => {
			const wrapper = shallow(<TextFilterEditor label="label" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when forArea is set', () => {
			const wrapper = shallow(<TextFilterEditor label="label" forArea />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('OldTextFilterEditor', () => {
		describe('rendering', () => {
			it('renders expected elements', () => {
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={['a']} placeholder="placeholder" validateSaveRef={{}} />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});

			it('renders expected elements when list is null', () => {
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={null} placeholder="placeholder" validateSaveRef={{}} />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});

			it('renders expected elements when list input requests suggestions text', () => {
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={['a']} placeholder="placeholder" validateSaveRef={{}} />
				);
				wrapper.find('#textFilterEditorInput').prop('getSuggestionText')({name: 'a'});
				expect(wrapper.getElement()).toMatchSnapshot();
			});
		});

		describe('behaviour', () => {
			it('calls onListChange when list is changed', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={['a']} validateSaveRef={{}} onListChange={onListChange} />
				);
				wrapper.find('#textFilterEditorInput').prop('onListChange')(['a']);
				expect(onListChange.mock.calls).toEqual([[['a']]]);
			});

			it('calls onListChange when list is empty', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={[]} validateSaveRef={{}} onListChange={onListChange} />
				);
				wrapper.find('#textFilterEditorInput').prop('onListChange')([]);
				expect(onListChange.mock.calls).toEqual([[null]]);
			});

			it('returns the suggestion name if the order is 2', () => {
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={[]} placeholder="placeholder" validateSaveRef={{}} />
				);
				expect(wrapper.find('ListInput').prop('getHighlightKey')({name: 'test', order: 2})).toBe('test');
			});

			it('returns null if the suggestion order is not 2', () => {
				const wrapper = shallow(
					<OldTextFilterEditor label="label" list={[]} placeholder="placeholder" validateSaveRef={{}} />
				);
				expect(wrapper.find('ListInput').prop('getHighlightKey')({name: 'test', order: 0})).toBeNull();
			});

			it('sets error when text is set', () => {
				const ref = {};
				shallow(<OldTextFilterEditor label="label" list={['a']} placeholder="placeholder" validateSaveRef={ref} />);
				expect(ref).not.toEqual({});
				expect(ref.current()).toBe(true);
			});

			it('does not set an error when no text is set', () => {
				const ref = {};
				const setError = jest.fn();
				useState.mockReturnValueOnce(['something']).mockReturnValueOnce([null, setError]);

				shallow(<OldTextFilterEditor label="label" list={['a']} placeholder="placeholder" validateSaveRef={ref} />);
				expect(ref).not.toEqual({});
				expect(ref.current()).toBe(false);
				expect(setError.mock.calls).toEqual([['Press Enter to add items to the list.']]);
			});
		});
	});

	describe('NewTextFilterEditor', () => {
		describe('rendering', () => {
			it('renders expected elements', () => {
				const wrapper = shallow(
					<NewTextFilterEditor id="test" label="label" list={['a']} placeholder="placeholder" buttonId="button" />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});

			it('renders expected elements when list is null', () => {
				const wrapper = shallow(
					<NewTextFilterEditor id="test" label="label" list={null} placeholder="placeholder" buttonId="button" />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});
		});

		describe('behaviour', () => {
			it('calls onSave when the editor calls onSave', () => {
				const list = ['a'];
				const onSave = jest.fn();
				const wrapper = shallow(
					<NewTextFilterEditor
						id="test"
						label="label"
						list={list}
						placeholder="placeholder"
						chipId="button"
						onSave={onSave}
					/>
				);
				wrapper.find('#test').prop('onSave')();
				expect(onSave.mock.calls).toEqual([[list]]);
			});

			it('calls onSave when the editor calls onSave and the list is empty', () => {
				const onSave = jest.fn();
				const wrapper = shallow(
					<NewTextFilterEditor
						id="test"
						label="label"
						list={[]}
						placeholder="placeholder"
						chipId="button"
						onSave={onSave}
					/>
				);
				wrapper.find('#test').prop('onSave')();
				expect(onSave.mock.calls).toEqual([[null]]);
			});

			it('calls setError when the editor calls onSave and text is not empty', () => {
				const onSave = jest.fn();
				const setError = jest.fn();
				useState.mockReturnValueOnce([[]]).mockReturnValueOnce(['foo']).mockReturnValueOnce([null, setError]);
				const wrapper = shallow(
					<NewTextFilterEditor
						id="test"
						label="label"
						list={['a']}
						placeholder="placeholder"
						chipId="button"
						onSave={onSave}
					/>
				);
				wrapper.find('#test').prop('onSave')();
				expect(onSave.mock.calls).toEqual([]);
				expect(setError.mock.calls).toEqual([['Press Enter to add items to the list.']]);
			});
		});
	});
});
