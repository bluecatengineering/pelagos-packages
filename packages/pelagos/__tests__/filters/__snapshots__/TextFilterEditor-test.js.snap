// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextFilterEditor NewTextFilterEditor rendering renders expected elements 1`] = `
<FilterEditor
  aria-labelledby="random-id-list-label"
  id="random-id"
  onSave={[Function]}
>
  <ListInput
    column={true}
    error={null}
    getHighlightKey={[Function]}
    getItemKey={[Function]}
    getItemName={[Function]}
    getSuggestionText={[Function]}
    id="random-id-list"
    label="label"
    list={
      [
        "a",
      ]
    }
    onErrorChange={[MockFunction]}
    onListChange={[MockFunction]}
    onTextChange={[MockFunction]}
    placeholder="placeholder"
    renderSuggestion={[MockFunction]}
    type="text"
    value=""
  />
</FilterEditor>
`;

exports[`TextFilterEditor NewTextFilterEditor rendering renders expected elements when list is null 1`] = `
<FilterEditor
  aria-labelledby="random-id-list-label"
  id="random-id"
  onSave={[Function]}
>
  <ListInput
    column={true}
    error={null}
    getHighlightKey={[Function]}
    getItemKey={[Function]}
    getItemName={[Function]}
    getSuggestionText={[Function]}
    id="random-id-list"
    label="label"
    list={[]}
    onErrorChange={[MockFunction]}
    onListChange={[MockFunction]}
    onTextChange={[MockFunction]}
    placeholder="placeholder"
    renderSuggestion={[MockFunction]}
    type="text"
    value=""
  />
</FilterEditor>
`;

exports[`TextFilterEditor OldTextFilterEditor rendering renders expected elements 1`] = `
<ListInput
  column={true}
  error={null}
  getHighlightKey={[Function]}
  getItemKey={[Function]}
  getItemName={[Function]}
  getSuggestionText={[Function]}
  id="textFilterEditorInput"
  label="label"
  list={
    [
      "a",
    ]
  }
  onErrorChange={[MockFunction]}
  onListChange={[Function]}
  onTextChange={[MockFunction]}
  placeholder="placeholder"
  renderSuggestion={[MockFunction]}
  type="text"
  value=""
/>
`;

exports[`TextFilterEditor OldTextFilterEditor rendering renders expected elements when list input requests suggestions text 1`] = `
<ListInput
  column={true}
  error={null}
  getHighlightKey={[Function]}
  getItemKey={[Function]}
  getItemName={[Function]}
  getSuggestionText={[Function]}
  id="textFilterEditorInput"
  label="label"
  list={
    [
      "a",
    ]
  }
  onErrorChange={[MockFunction]}
  onListChange={[Function]}
  onTextChange={[MockFunction]}
  placeholder="placeholder"
  renderSuggestion={[MockFunction]}
  type="text"
  value=""
/>
`;

exports[`TextFilterEditor OldTextFilterEditor rendering renders expected elements when list is null 1`] = `
<ListInput
  column={true}
  error={null}
  getHighlightKey={[Function]}
  getItemKey={[Function]}
  getItemName={[Function]}
  getSuggestionText={[Function]}
  id="textFilterEditorInput"
  label="label"
  list={[]}
  onErrorChange={[MockFunction]}
  onListChange={[Function]}
  onTextChange={[MockFunction]}
  placeholder="placeholder"
  renderSuggestion={[MockFunction]}
  type="text"
  value=""
/>
`;

exports[`TextFilterEditor rendering renders expected elements when forArea is not set 1`] = `
<OldTextFilterEditor
  label="label"
/>
`;

exports[`TextFilterEditor rendering renders expected elements when forArea is set 1`] = `
<NewTextFilterEditor
  label="label"
/>
`;
